types/publish.mjs

```diff
@@ -0,0 +1,445 @@
+  This module is generated automatically as part of Ember's publishing process and
+  should never be edited manually.
+
+  To use these type definitions, add this import to any TypeScript file in your
+  Ember app or addon:
+
+  \`\`\`ts
+  import 'ember-source/types';
+  import 'ember-source/types/preview';
+  \`\`\`
+
+  @module
+ */
+
+// This works because each of these modules presents \`declare module\` definition
+// of the module and *only* that, so importing this file in turn makes those
+// module declarations "visible" automatically throughout a consuming project.
+// Combined with use of \`typesVersions\` (or, in the future, possibly \`exports\`)
+// in \`package.json\`, this allows users to import the types without knowing the
+// exact layout details.
+//
+// Somewhat annoyingly, every single module in the graph must appear here. For
+// now, while we are publishing ambient types, that means we must maintain this
+// by hand. When we start emitting types from the source, we will need to do the
+// same work, but automatically.
+
+// STATUS NOTE: this does not yet include Ember's full public API, only the
+// subset of it for which we have determined the types are ready to stabilize.
+//
+// Over time, it will come to include *all* of Ember's types, and the matching
+// \`preview\` types will become empty. This is means that someone who writes the
+// import we recommend--
+//
+// \`\`\`ts
+// import 'ember-source/types';
+// import 'ember-source/types/preview';
+// \`\`\`
+//
+// --will always get the most up-to-date mix of preview and stable types, with
+// no extra effort required.
+
+${MODULES_PLACEHOLDER}
+`;
+
+const TYPES_DIR = path.join('types', 'stable');
+
+async function main() {
+  fs.rmSync(TYPES_DIR, { recursive: true, force: true });
+  fs.mkdirSync(TYPES_DIR, { recursive: true });
+
+  spawnSync('yarn', ['tsc', '--project', 'tsconfig/publish-types.json']);
+
+  // This is rooted in the `TYPES_DIR` so that the result is just the names of
+  // the modules, as generated directly from the tsconfig above.
+  let moduleNames = glob.sync('**/*.d.ts', {
+    ignore: 'index.d.ts', // ignore the root file itself if it somehow exists
+    cwd: TYPES_DIR,
+  });
+
+  for (let moduleName of moduleNames) {
+    wrapInDeclareModule(moduleName);
+  }
+
+  let sideEffectModules = moduleNames
+    .filter((module) => !PREVIEW_MODULES.includes(module))
+    .map((moduleName) => `import './${moduleName}';`)
+    .join('\n');
+
+  let stableIndexDTsContents = BASE_INDEX_D_TS.replace(MODULES_PLACEHOLDER, sideEffectModules);
+  fs.writeFileSync(path.join(TYPES_DIR, 'index.d.ts'), stableIndexDTsContents);
+}
+
+/**
+ * @param {string} moduleName
+ */
+function wrapInDeclareModule(moduleName) {
+  let modulePath = path.join(TYPES_DIR, moduleName);
+
+  /** @type {string} */
+  let contents;
+  try {
+    contents = fs.readFileSync(modulePath, { encoding: 'utf-8' });
+  } catch (e) {
+    console.error(`Error reading ${modulePath}: ${e}`);
```

Comment:
Do we care about exiting with a nonzero code if something like this happens?

Reply:
Ah, good point. Yes, we should; I'll push up a change to that effect.